pype/lib.py

```diff
@@ -1386,3 +1391,76 @@ def ffprobe_streams(path_to_file):
     popen_output = popen.communicate()[0]
     log.debug("FFprobe output: {}".format(popen_output))
     return json.loads(popen_output)["streams"]
+
+
+def decompose_url(url):
+    components = {
+        "scheme": None,
+        "host": None,
+        "port": None,
+        "username": None,
+        "password": None,
+        "auth_db": ""
+    }
+
+    result = urlparse(url)
+
+    components["scheme"] = result.scheme
+    components["host"] = result.hostname
+    try:
+        components["port"] = result.port
+    except ValueError:
+        raise RuntimeError("invalid port specified")
+    components["username"] = result.username
+    components["password"] = result.password
+
+    try:
+        components["auth_db"] = parse_qs(result.query)['authSource'][0]
+    except KeyError:
+        # no auth db provided, mongo will use the one we are connecting to
+        pass
+
+    return components
+
+
+def compose_url(scheme=None,
+                host=None,
+                username=None,
+                password=None,
+                database=None,
+                collection=None,
+                port=None,
+                auth_db=""):
+
+    url = "{scheme}://"
+
+    if username and password:
+        url += "{username}:{password}@"
+
+    url += "{host}"
+
+    if database:
+        url += "/{database}"
+
+    if database and collection:
+        url += "/{collection}"
+
+    if port:
+        url += ":{port}"
+
+    url += auth_db
```

Comment:
It should be
```
if auth_db:
    url += "?authSource={}".format(auth_db)
```
But we maybe should store whole url query in decompose instead of just storing one key from it. So on lines 1417-1421 won't be try->except but just
```
components["query"] = result.query
```
and here
```
if query:
    url += "?{}".format(query)
```

Reply:
Sure, done.